subsys/bluetooth/lib/ead.c

```diff
@@ -0,0 +1,221 @@
+	 * of the CCM nonce shall be set to the most significant bit of the Randomizer field.
+	 */
+	randomizer[4] |= 1 << BT_EAD_RANDOMIZER_DIRECTION_BIT;
+
+	return 0;
+}
+
+static int bt_ead_generate_nonce(const uint8_t *iv, const uint8_t *randomizer, uint8_t *nonce)
+{
+	int err;
+	uint8_t new_randomizer[BT_EAD_RANDOMIZER_SIZE];
+
+	if (randomizer == NULL) {
+		err = bt_ead_generate_randomizer(new_randomizer);
+
+		if (err != 0) {
+			LOG_ERR("Failed to generate Randomizer");
+			return -ECANCELED;
+		}
+
+		randomizer = new_randomizer;
+	}
+
+	memcpy(&nonce[0], randomizer, BT_EAD_RANDOMIZER_SIZE);
+	memcpy(&nonce[BT_EAD_RANDOMIZER_SIZE], iv, BT_EAD_IV_SIZE);
+
+	return 0;
+}
+
+size_t bt_ead_get_encrypted_data_size(const struct bt_data *payload, size_t num_elts)
+{
+	size_t ead_size = 0;
+
+	for (size_t i = 0; i < num_elts; i++) {
+		ead_size +=
+			sizeof(payload[i].data_len) + sizeof(payload[i].type) + payload[i].data_len;
+	}
+
+	ead_size += BT_EAD_RANDOMIZER_SIZE + BT_EAD_MIC_SIZE;
+
+	return ead_size;
+}
+
+size_t bt_ead_get_decrypted_data_size(const struct bt_data *payload)
+{
+	if (payload->type != BT_DATA_ENCRYPTED_AD_DATA) {
+		return 0;
+	}
+
+	size_t decrypted_data_size;
+
+	decrypted_data_size = (payload->data_len - BT_EAD_RANDOMIZER_SIZE) - BT_EAD_MIC_SIZE;
+
+	return decrypted_data_size;
+}
+
+size_t bt_data_to_ad_structure(const struct bt_data *input, uint8_t *output_ad_structure)
+{
+	uint8_t ad_data_len = input->data_len;
+	uint8_t data_len = ad_data_len + 1;
+
+	output_ad_structure[0] = data_len;
+	output_ad_structure[1] = input->type;
+
+	memcpy(&output_ad_structure[2], input->data, ad_data_len);
+
+	return data_len + 1;
+}
+
+int ead_encrypt(const uint8_t *session_key, const uint8_t *iv, const uint8_t *randomizer,
+		const uint8_t *payload, size_t payload_size, uint8_t *encrypted_payload)
+{
+	int err;
+	uint8_t nonce[BT_EAD_NONCE_SIZE];
+	size_t ead_size = BT_EAD_RANDOMIZER_SIZE + payload_size + BT_EAD_MIC_SIZE;
+
+	err = bt_ead_generate_nonce(iv, randomizer, nonce);
+
+	if (err != 0) {
+		return -ECANCELED;
+	}
+
+	memcpy(encrypted_payload, nonce, BT_EAD_RANDOMIZER_SIZE);
+
+	err = bt_ccm_encrypt(session_key, nonce, payload, payload_size, bt_ead_aad, BT_EAD_AAD_SIZE,
+			     &encrypted_payload[BT_EAD_RANDOMIZER_SIZE], BT_EAD_MIC_SIZE);
+
+	LOG_HEXDUMP_DBG(encrypted_payload, ead_size, "Encrypted Data: ");
+
+	if (err != 0) {
+		LOG_ERR("Failed to encrypt the payload (bt_ccm_encrypt err %d)", err);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+int bt_ead_encrypt(const uint8_t *session_key, const uint8_t *iv, const uint8_t *payload,
+		   size_t payload_size, uint8_t *encrypted_payload)
+{
+	if (session_key == NULL || iv == NULL || payload == NULL || encrypted_payload == NULL) {
+		LOG_ERR("At least one of the argument is a NULL pointer.");
+		return -EINVAL;
+	}
+
+	if (payload_size == 0) {
+		LOG_WRN("payload_size is set to 0. The encrypted result will be filled with only "
+			"the Randomizer and the MIC.");
+	}
```

Comment:
Is it really valid to encrypt 0 octets of data?

Reply:
Nothing is prohibiting it. Not in the RFC of the CCM nor the CCM description in the Bluetooth Core.
I tried it and it will just create a Randomizer with a MIC. The MIC will be computed using the nonce and the AAD and so will be valid.